src/storage-client/src/controller.rs

```diff
@@ -2405,6 +2439,110 @@ where
         let json_state = serde_json::to_value(shard_state)?;
         Ok(json_state)
     }
+
+    async fn record_frontiers(&mut self, mut frontiers: BTreeMap<GlobalId, Self::Timestamp>) {
+        let old_rows = &mut self.frontiers_rows;
+        let mut updates = Vec::with_capacity(old_rows.len());
+
+        // Retract existing rows.
+        updates.extend(old_rows.drain(..).map(|row| (row, -1)));
+
+        let mut push_frontier = |object_id: GlobalId, time: T| {
+            let row = Row::pack_slice(&[
+                Datum::String(&object_id.to_string()),
+                Datum::MzTimestamp(mz_repr::Timestamp::from_bytes(time.encode())),
+            ]);
+            updates.push((row.clone(), 1));
+            old_rows.push(row);
+        };
+
+        // Add the frontiers known to the storage controller.
+        for (collection_id, collection) in &self.state.collections {
+            // Some of the storage frontiers are not advanced by us but by our users, which might
```

Comment:
The one place I am aware of is where the compute controller tells storage about the frontiers of MV collections: https://github.com/MaterializeInc/materialize/blob/eabd060c47134d2359eeb3e28c0afaa51efb6c51/src/compute-client/src/controller/instance.rs#L923-L937C55.

> why are the storage frontiers preferred over the ones passed to us?

Probably debatable, but my assumption was that the storage controller should be the authoritative source about storage frontiers. Maybe it's worth asserting here that duplicate frontiers always match. That must be the case today, but you could imagine a future in which the compute controller doesn't tell storage about new MV frontiers immediately, or storage learns about new MV frontiers before the compute controller does.

Reply:
What I was thinking was that maybe the contract should be that `frontiers` only contains foreign objects (indexes) so that the storage controller can augment it with its own state and assert that they don't overlap

Reply:
Fine by me. At one point we have to filter out the storage objects from the compute frontiers. We can do that one level higher in the `Controller` if you prefer. I'd prefer to not do it in the compute controller, because then we'd have to change the name for the method that retrieves these frontiers to something awkward like `write_frontiers_filtered_by_storage_objects`.

Reply:
Turns out this doesn't actually work: For MVs (the only frontiered objects shared between compute and storage currently), the compute controller knows the per-replica frontiers, while the storage controller only knows the global frontiers. Since we want to show per-replica frontiers for MV, we need to prefer external frontiers before storage frontiers in this method.